crates/optimism/payload/src/builder.rs

```diff
@@ -542,3 +554,84 @@ where
 
     Ok(BuildOutcome::Better { payload, cached_reads })
 }
+
+#[derive(Debug, thiserror::Error)]
+// Error type for EIP-1559 parameters
+pub enum EIP1559ParamError {
+    #[error("No EIP-1559 parameters provided")]
+    /// No EIP-1559 parameters provided
+    NoEIP1559Params,
+    #[error("Invalid elasticity parameter")]
+    /// Invalid denominator parameter
+    InvalidDenominator,
+    #[error("Invalid denominator parameter")]
+    /// Invalid elasticity parameter
+    InvalidElasticity,
+}
+
+/// Extracts the Holcene 1599 parameters from the encoded form:
+/// <https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/holocene/exec-engine.md#eip1559params-encoding>
+pub fn decode_eip_1559_params(eip_1559_params: B64) -> Result<(u32, u32), EIP1559ParamError> {
+    let elasticity: [u8; 4] =
+        eip_1559_params.0[..4].try_into().map_err(|_| EIP1559ParamError::InvalidElasticity)?;
+
+    let denominator: [u8; 4] =
+        eip_1559_params.0[4..8].try_into().map_err(|_| EIP1559ParamError::InvalidDenominator)?;
+
+    Ok((u32::from_be_bytes(elasticity), u32::from_be_bytes(denominator)))
+}
+
+fn get_holocene_extra_data(
+    attributes: &OptimismPayloadBuilderAttributes,
+    default_base_fee_params: BaseFeeParams,
+) -> Result<Bytes, EIP1559ParamError> {
+    let eip_1559_params = attributes.eip_1559_params.ok_or(EIP1559ParamError::NoEIP1559Params)?;
+
+    let mut extra_data = [0u8; 9];
+    // If eip 1559 params aren't set, use the canyon base fee param constants
+    // otherwise use them
+    if eip_1559_params == B64::ZERO {
+        extra_data[1..5].copy_from_slice(
+            &(default_base_fee_params.max_change_denominator as u32).to_be_bytes(),
+        );
+        extra_data[5..9]
+            .copy_from_slice(&(default_base_fee_params.elasticity_multiplier as u32).to_be_bytes());
```

Comment:
Is it possible that the `as` cast here fails? If so can we add an error variant?